### PR TITLE
Fix multi-dimensional `VectorOfArray` broadcast

### PR DESCRIPTION
**Pull request by jlchan**
Addresses https://github.com/SciML/RecursiveArrayTools.jl/issues/373

The fix just involves modifying `Base.copy` for `VectorOfArray` with multi-dimensional parent arrays. This could replace the old `Base.copy` implementation, but I left the old implementation there just in case it changes behavior upstream.

### CODE REVIEW

File: src/vector_of_array.jl [hunk 1]
```diff
@@ -840,12 +840,37 @@ end
...
+        # this is the default behavior in v3.15.0
+        N = narrays(bc)
+        return VectorOfArray(map(1:N) do i
+            copy(unpack_voa(bc, i))
+        end)
+    else # if parent isa AbstractArray            
+        return VectorOfArray(map(enumerate(Iterators.product(axes(parent)...))) do (i, _)
+            copy(unpack_voa(bc, i))
+        end)
+    end
```
Comment (jlchan) [inline src/vector_of_array.jl:873]: Side note, this entire block can be replaced by the more general second branch 
```julia
return VectorOfArray(map(enumerate(Iterators.product(axes(parent)...))) do (i, _)
    copy(unpack_voa(bc, i))
end)
```
since `enumerate(Iterators.product(axes(parent)...))) do (i, _)` basically recovers `map(1:N) do i` for an `AbstractVector`. 

I left the old behavior in since I wasn't sure if it would break something upstream or change performance.

File: src/vector_of_array.jl [hunk 2]
```diff
@@ -719,7 +718,7 @@ end
 # for VectorOfArray with multi-dimensional parent arrays of arrays where all elements are the same type
 function Base.similar(vec::VectorOfArray{
         T, N, AT}) where {T, N, AT <: AbstractArray{<:AbstractArray{T}}}
-    return VectorOfArray(similar(Base.parent(vec)))
+    return VectorOfArray(similar.(Base.parent(vec)))
```
Comment (jlchan) [inline src/vector_of_array.jl:721]: Broadcasting `similar` here avoids some `#undef` issues. 

Side note: this implementation could replace these 3 functions https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/src/vector_of_array.jl#L715-L729

However, doing so breaks these tests because `similar` can't be called on `<:Real` values: https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/test/basic_indexing.jl#L221-L229
These tests don't seem correct to me; the parent array `x` isn't a recursive array (`eltype(x)` is not `<:AbstractArray`), so to me it doesn't make sense to test this broadcast edge case. Maybe I'm missing something though.

**Comment by jlchan** [comment]
Applying formatting introduces a bunch of spurious file changes so I'm skipping it for now.

File: src/vector_of_array.jl (hunk 2 above)
Comment (jlchan) [reply src/vector_of_array.jl:721]: Side note: this implementation could replace these 3 functions https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/src/vector_of_array.jl#L715-L729

However, doing so breaks these tests because `similar` can't be called on `<:Real` values: https://github.com/SciML/RecursiveArrayTools.jl/blob/eacbe3fa275440ea863db85e63d627dd103205d6/test/basic_indexing.jl#L221-L229
These tests don't seem correct to me; the parent array `x` doesn't have underlying data structure `Vector{AbstractArray{T}}`, so it shouldn't be a valid edge case. Maybe I'm missing something?

Comment (ChrisRackauckas) [reply src/vector_of_array.jl:721]: Numbers are also allowed. That's used a lot downstream.

Comment (ChrisRackauckas) [reply src/vector_of_array.jl:721]: We can special-case that case where the elements are numbers, but we cannot throw it out.

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by jlchan** [comment]
Thanks for the quick review!
